lib/src/git.rs

```diff
@@ -1905,3 +2232,512 @@ pub fn parse_gitmodules(
+            "external git program failed with non-utf8 output: {e:?}\n--- stderr ---\n{lossy_err}",
+        ))
+    })?;
+
+    // There are some git errors we want to parse out
+
+    // GitFetchError::NoSuchRemote
+    //
+    // To say this, git prints out a lot of things, but the first line is of the
+    // form:
+    // `fatal: '<branch_name>' does not appear to be a git repository`
+    let git_err = {
+        let mut git_err = None;
+        if let Some(first_line) = stderr.lines().next() {
+            if first_line.starts_with("fatal: '")
+                && first_line.ends_with("' does not appear to be a git repository")
+            {
+                let mut split = first_line.split('\'');
+                split.next(); // ignore prefix
+                let branch_name = split.next();
+                if let Some(bname) = branch_name {
+                    git_err = Some(Err(GitFetchError::NoSuchRemote(bname.to_string())));
+                }
+            }
+        }
+
+        git_err
+    };
+    if let Some(e) = git_err {
+        return e;
+    }
+
+    // even though --prune is specified, if a particular refspec is asked for but
+    // not present in the remote, git will error out.
+    // we ignore it explicitly
+    //
+    // The first line is of the form:
+    // `fatal: couldn't find remote ref refs/heads/<ref>`
+    if let Some(first_line) = stderr.lines().next() {
+        if first_line.starts_with("fatal: couldn't find remote ref refs/heads/") {
+            let mut sp = first_line.split("refs/heads/");
+            sp.next();
+            prunes.push(sp.next().unwrap().to_string());
+            return Ok(output);
+        }
+    }
+
+    Err(GitFetchError::ExternalGitError(format!(
+        "external git program failed:\n{stderr}",
+    )))
+}
+
+fn fork_git_fetch(
+    git_dir: &str,
+    work_tree: &str,
+    remote_name: &str,
+    depth: Option<NonZeroU32>,
+    refspec: &str,
+    prunes: &mut Vec<String>,
+) -> Result<(), GitFetchError> {
+    let depth_arg = depth.map(|x| format!("--depth={x}"));
+    tracing::debug!(
+        "shelling out to `git --git-dir={} --work-tree={} fetch --prune{} {} {}",
+        git_dir,
+        work_tree,
+        depth_arg
+            .as_ref()
+            .map(|x| format!(" {x}"))
+            .unwrap_or("".to_string()),
+        remote_name,
+        refspec
+    );
+
+    let remote_git = {
+        let mut cmd = Command::new("git");
+
+        cmd.arg(format!("--git-dir={git_dir}"))
+            .arg(format!("--work-tree={work_tree}"))
+            .arg("fetch")
+            .arg("--prune");
+
+        if let Some(depth) = depth_arg {
+            cmd.arg(depth);
+        }
+
+        cmd.arg(remote_name)
+            .arg(refspec)
+            .stdin(Stdio::null())
+            .stderr(Stdio::piped())
+            .spawn()
+    }?;
+
+    let output = remote_git.wait_with_output()?;
+    let _ = convert_git_fetch_output_to_fetch_result(output, prunes)?;
```

Comment:
suggestion: I would annotate the type of the result variable here (or use the result), as it can catch issues when the return value of the function changes and now it's meant to be used.